packages/typescript-group-checker/src/compilers/compiler-with-watch.ts

```diff
@@ -0,0 +1,154 @@
+
+  private readonly allTSConfigFiles: Set<string>;
+  private readonly tsconfigFile: string;
+  private currentTask = new Task();
+  private currentErrors: ts.Diagnostic[] = [];
+  private readonly sourceFiles: DependencyFile[] = [];
+
+  constructor(private readonly log: Logger, private readonly options: StrykerOptions, private readonly fs: MemoryFileSystem) {
+    this.tsconfigFile = toPosixFileName(this.options.tsconfigFile);
+    this.allTSConfigFiles = new Set([path.resolve(this.tsconfigFile)]);
+  }
+
+  public async init(): Promise<{ dependencyFiles: DependencyFile[]; errors: ts.Diagnostic[] }> {
+    guardTSVersion();
+    this.guardTSConfigFileExists();
+    const buildModeEnabled = determineBuildModeEnabled(this.tsconfigFile);
+
+    const host = ts.createSolutionBuilderWithWatchHost(
+      {
+        ...ts.sys,
+        readFile: (fileName) => {
+          const content = this.fs.getFile(fileName)?.content;
+          if (content && this.allTSConfigFiles.has(path.resolve(fileName))) {
+            return this.adjustTSConfigFile(fileName, content, buildModeEnabled);
+          }
+          return content;
+        },
+        getModifiedTime: (pathName: string) => {
+          return this.fs.getFile(pathName)?.modifiedTime;
+        },
+        watchFile: (filePath: string, callback: ts.FileWatcherCallback) => {
+          const file = this.fs.getFile(filePath);
+          if (file) {
+            file.watcher = callback;
+          }
+
+          return {
+            close: () => {
+              delete this.fs.getFile(filePath)!.watcher;
+            },
+          };
+        },
+        writeFile: (fileName, data) => {
+          this.fs.writeFile(fileName, data);
+        },
+      },
+      undefined,
+      (error) => {
+        this.currentErrors.push(error);
+      },
+      (status) => {
+        this.log.debug(status.messageText.toString());
+      },
+      (summary) => {
+        this.currentTask.resolve();
+      }
+    );
+
+    host.afterProgramEmitAndDiagnostics = (program) => {
+      const sourceFiles = program.getSourceFiles().map((f: any) => {
+        const file: { fileName: string; imports: Array<{ text: string }> } = f;
+
+        const result = {
+          fileName: this.resolveFilename(file.fileName),
+          imports: new Set(
+            file.imports
+              .filter((i) => {
+                if (i.text) {
+                  return true;
+                } else {
+                  return false;
+                }
+              })
+              .map((i) => ts.resolveModuleName(i.text, file.fileName, {}, host).resolvedModule?.resolvedFileName ?? `${i.text}-not_resolved`)
+              .map((i) => this.resolveFilename(i))
+          ),
+        };
+
+        return result;
+      });
+
+      for (const sourceFile of sourceFiles) {
+        const indexInSourceFileArray = this.sourceFiles.findIndex((sourceFileThis) => sourceFileThis.fileName === sourceFile.fileName);
```

Comment:
Een array heeft ook een `find` methode die het element direct teruggeeft (als er een item is)